lib/content_encoding.c

```diff
@@ -710,6 +714,93 @@ static const struct content_encoding brotli_encoding = {
 #endif
 
 
+#ifdef HAVE_ZSTD
+/* Writer parameters. */
+struct zstd_params {
+  ZSTD_DStream *zds;    /* State structure for zstd. */
+  void *decomp;
+};
+
+static CURLcode zstd_init_writer(struct connectdata *conn,
+                                 struct contenc_writer *writer)
+{
+  struct zstd_params *zp = (struct zstd_params *)&writer->params;
+  (void)conn;
+
+  if(!writer->downstream)
+    return CURLE_WRITE_ERROR;
+
+  zp->zds = ZSTD_createDStream();
+  zp->decomp = NULL;
+  return zp->zds ? CURLE_OK : CURLE_OUT_OF_MEMORY;
+}
+
+static CURLcode zstd_unencode_write(struct connectdata *conn,
+    struct contenc_writer *writer,
+    const char *buf, size_t nbytes)
+{
+  CURLcode result = CURLE_OK;
+  struct zstd_params *zp = (struct zstd_params *)&writer->params;
+  ZSTD_inBuffer in;
+  ZSTD_outBuffer out;
+  size_t errorCode;
+
+  if(!zp->decomp) {
+    zp->decomp = malloc(DSIZ);
```

Comment:
I'm curious on this malloc, and maybe @felixhandte can help us out here (I tried to find docs that would explain this but I failed). How is this particular memory block size chosen? Is there a benefit to use any particular size and if so, what's the trade-offs? `DSIZ` is `CURL_MAX_WRITE_SIZE`, 16K by default.

Reply:
this is size used by other decompressor (both brotli and zlib). I believed this was optimal size for Curl_unencode_write.
Zstd is happy with this size or other size.

Reply:
16KB should be fine here.

Optimizing output buffer sizes for Zstd decompression is mostly a best effort sort of thing, and generally not performance critical. If Zstd gets a frame that it knows will be smaller than the provided output buffer, it can decompress straight into that buffer. But if you receive a streamed response (i.e., one that doesn't announce its total size in the [`Frame_Content_Size` field](https://tools.ietf.org/html/rfc8478#section-3.1.1.1.1.1) in the frame header), Zstd has to internally buffer the output anyways to maintain its history window. So the output buffer size doesn't really matter, we're just `memcpy`-ing into it from that internal buffer that we are actually doing the decompression into.

Which, by the way, raises the topic of window sizes. `libzstd` has a default limit of [128 MB](https://github.com/facebook/zstd/blob/dev/lib/zstd.h#L1071-L1075), which means that the `ZSTD_DCtx` may internally allocate up to ~128 MB, and anything that requires more will be rejected. This limit [can be configured](https://github.com/facebook/zstd/blob/dev/lib/zstd.h#L515-L520): I'm not sure whether curl wants to expose this to users, or whether curl wants to select a different limit by default. The [RFC recommends for interoperability](https://tools.ietf.org/html/rfc8478#section-3.1.1.1.2) that, in the absence of negotiation, encoders not use a window larger than 8 MB:

> For improved interoperability, it's recommended for decoders to
> support values of Window_Size up to 8 MB and for encoders not to
> generate frames requiring a Window_Size larger than 8 MB.  It's
> merely a recommendation though, and decoders are free to support
> larger or lower limits, depending on local limitations.

Reply:
I made minor decompression code to follow @Cyan4973 suggestion about exit loop condition